h2o-core/src/main/java/water/H2O.java

```diff
@@ -2153,12 +2153,15 @@ public static void setFlatfile(HashSet<H2ONode> nodes) {
     return (HashSet<H2ONode>) STATIC_H2OS.clone();
   }
 
-  public static H2ONode reportClient(H2ONode client){
-    H2ONode oldClient = CLIENTS_MAP.put(client.getIpPortString(), client);
-    if(oldClient == null){
-      Log.info("New client discovered at " + client);
+  public static void reportClient(H2ONode client){
+    // consider only relevant clients
+    if(client._heartbeat._cloud_name_hash == H2O.SELF._heartbeat._cloud_name_hash){
```

Comment:
I am not sure if this is the best place to put it, better place is the communication layer. In my opinion this method should not even be called if there is such mismatch.

For example in UDPClientEvent.call() we acquire the Heartbeat and then call Paxos.doHeartbeat without checking the hash first. This can shutdown the cluster as well if the heartbeat is coming from an incompatible version of H2O (unless check jar is disabled).

I think the heartbeat should be immediately discarded if we detect it is coming from a different cluster. Same logic for CleintEvents.

Reply:
👍 agree, we can push it down - we can ignore messages coming from wrong clouds (in both modes multi/flatfile) .

However, i should mention the original motivation - it was to detect incorrect deployments and fail early. 
But i think we do not need that anymore? WDYT?

Reply:
So changed:

In `Paxos` `doHeartbeat` method we already have the check which ignores the messages from different cloud. The problem was that later in the method it broadcasts the client via the ClientEvent which is processed via different entry point and this can affect different clusters. The change was just to put the check for cloud name hash to `UDPClientEvent` call as well

Also ensured we ignore messages from different clouds in multicast mode.